File: roster.py
"""名簿の操作。画面からは編集できないため、必要なときは Pi 上でこれを使う。

    .venv/bin/python roster.py list
    .venv/bin/python roster.py add そら
    .venv/bin/python roster.py hide そら
    .venv/bin/python roster.py show そら
"""

import sys

from app import config, db


def main(argv: list[str]) -> int:
    if len(argv) < 2 or argv[1] not in ("list", "add", "hide", "show"):
        print(__doc__)
        return 1

    command = argv[1]
    conn = db.connect(config.DEFAULT_DB_PATH)
    db.init_schema(conn)

    if command == "list":
        for member in db.list_members(conn, include_hidden=True):
            mark = "" if member["active"] else "  (非表示)"
            print(f"{member['id']:3}  {member['name']}{mark}")
        return 0

    if len(argv) < 3:
        print("名前が要る")
        return 1
    name = argv[2]

    if command == "add":
        try:
            print(f"追加した: {db.add_member(conn, name)}  {name}")
        except ValueError as exc:
            print(exc)
            return 1
        return 0

    target = next(
        (m for m in db.list_members(conn, include_hidden=True) if m["name"] == name), None
    )
    if target is None:
        print(f"見つからない: {name}")
        return 1
    db.set_member_active(conn, target["id"], command == "show")
    print(f"{'表示' if command == 'show' else '非表示'}にした: {name}")
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: roster.js
/**
 * 名簿の操作。画面からは編集できないため、必要なときは Pi 上でこれを使う。
 *
 *     node roster.js list
 *     node roster.js add そら
 *     node roster.js hide そら
 *     node roster.js show そら
 */

import { DEFAULT_DB_PATH } from "./src/config.js";
import * as db from "./src/db.js";

const USAGE = [
  "node roster.js list",
  "node roster.js add <名前>",
  "node roster.js hide <名前>",
  "node roster.js show <名前>",
].join("\n");

export function main(argv) {
  const command = argv[0];
  if (!["list", "add", "hide", "show"].includes(command)) {
    console.log(USAGE);
    return 1;
  }

  const database = db.connect(process.env.DB_PATH ?? DEFAULT_DB_PATH);
  db.initSchema(database);

  try {
    if (command === "list") {
      for (const member of db.listMembers(database, true)) {
        const mark = member.active ? "" : "  (非表示)";
        console.log(`${String(member.id).padStart(3)}  ${member.name}${mark}`);
      }
      return 0;
    }

    const name = argv[1];
    if (!name) {
      console.log("名前が要る");
      return 1;
    }

    if (command === "add") {
      try {
        console.log(`追加した: ${db.addMember(database, name)}  ${name}`);
      } catch (error) {
        if (!(error instanceof db.InvalidName)) throw error;
        console.log(error.message);
        return 1;
      }
      return 0;
    }

    const target = db.listMembers(database, true).find((member) => member.name === name);
    if (target === undefined) {
      console.log(`見つからない: ${name}`);
      return 1;
    }
    db.setMemberActive(database, target.id, command === "show");
    console.log(`${command === "show" ? "表示" : "非表示"}にした: ${name}`);
    return 0;
  } finally {
    database.close();
  }
}

process.exitCode = main(process.argv.slice(2));
